python/sanlock.c

```diff
@@ -181,31 +182,33 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
 
     for (i = 0; i < num_disks; i++) {
         PyObject *disk;
-        const char *path;
         uint64_t offset;
 
         disk = PyList_GetItem(obj, i);
+        path = NULL;
 
         if (!PyTuple_Check(disk)) {
             set_error(PyExc_ValueError, "Invalid disk %s", disk);
             goto exit_fail;
 
         }
 
-        if (!PyArg_ParseTuple(disk, "sK", &path, &offset)) {
+        if (!PyArg_ParseTuple(disk, "O&K", pypath_converter, &path, &offset)) {
             /* Override the error since it confusing in this context. */
             set_error(PyExc_ValueError, "Invalid disk %s", disk);
             goto exit_fail;
         }
 
-        strncpy(res->disks[i].path, path, SANLK_PATH_LEN - 1);
+        strncpy(res->disks[i].path, PyBytes_AsString(path), SANLK_PATH_LEN - 1);
+        Py_DECREF(path);
         res->disks[i].offset = offset;
     }
 
     *res_ret = res;
     return 0;
 
 exit_fail:
+    Py_XDECREF(path);
```

Comment:
Cleaning up path both globally and both once per loop is complicated and fragile. It is better to extract the code for parsing single (path, offset) item to a helper function that will hide the temporary objects  and have simple cleanup.

So should have something like:
```
parse_disks()
{
    allocate res...

    for (i = 0; i < num_disks; i++) {
        disk = PyList_GetItem(obj, i);
        if (parse_disks_item(disk, &res->disks[i]))
            goto exit_fail;
    }
    
exit_fail:
    ...
}
```